src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -293,3 +290,23 @@
       wrap-with-certificate-cn
       wrap-with-default-body
       wrap-with-debug-logging))
+
+(defn wrap-with-parent-check
+  "Middleware that checks the parent exists before serving the rest of the
+   application. This ensures we always return 404's on child paths when the
+   parent data is empty."
+  [app version parent id]
+  (fn [{:keys [globals] :as req}]
+    (let [{:keys [scf-read-db url-prefix]} globals
+          status-fn (case parent
```

Comment:
I guess what bothers me is that the existence checks should be visible when I'm looking around the code for an endpoint, and right now its kind of buried in that endpoints middleware.